NitroxClient/GameLogic/ChatUI/PlayerChatManager.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System.Collections;
+using NitroxClient.Communication.Abstract;
+using NitroxClient.MonoBehaviours.Gui.Chat;
+using NitroxClient.Unity.Helper;
+using NitroxModel.Packets;
+using UnityEngine;
+
+namespace NitroxClient.GameLogic.ChatUI
+{
+    public class PlayerChatManager
+    {
+        public PlayerChatManager(IMultiplayerSession multiplayerSession)
+        {
+            this.multiplayerSession = multiplayerSession;
+            Player.main.StartCoroutine(LoadChatLogAsset());
+        }
+
+        private PlayerChat playerChat;
+        public Transform PlayerChaTransform => playerChat.transform;
+        private readonly IMultiplayerSession multiplayerSession;
+
+        public void ShowChat() => Player.main.StartCoroutine(ShowChatAsync());
+        private IEnumerator ShowChatAsync()
+        {
+            while (PlayerChat.IsLoading)
+            {
+                yield return null;
+            }
+            playerChat.Show();
+        }
+
+        public void HideChat() => Player.main.StartCoroutine(HideChatAsync());
+        private IEnumerator HideChatAsync()
+        {
+            while (PlayerChat.IsLoading)
+            {
+                yield return null;
+            }
+            playerChat.Deselect();
+            playerChat.Hide();
+        }
+
+        public void SelectChat() => Player.main.StartCoroutine(SelectChatAsync());
+        private IEnumerator SelectChatAsync()
+        {
+            while (PlayerChat.IsLoading)
+            {
+                yield return null;
+            }
+            playerChat.Show();
+            playerChat.Select();
+        }
+
+        public void AddMessage(string playerName, string message, Color color) => Player.main.StartCoroutine(AddMessageAsync(playerName, message, color));
+        private IEnumerator AddMessageAsync(string playerName, string message, Color color)
+        {
+            while (PlayerChat.IsLoading)
+            {
+                yield return null;
+            }
+
+            playerChat.WriteLogEntry(playerName, message, color);
+        }
+
+        public void SendMessage()
+        {
+            if (playerChat.inputText.Trim() != "")
+            {
+                multiplayerSession.Send(new ChatMessage(multiplayerSession.Reservation.PlayerId, playerChat.inputText));
+                playerChat.WriteLogEntry(multiplayerSession.AuthenticationContext.Username, playerChat.inputText, multiplayerSession.PlayerSettings.PlayerColor);
+                playerChat.inputText = "";
+            }
+            playerChat.Select();
+        }
+
+        private IEnumerator LoadChatLogAsset()
+        {
+            CoroutineWithData coroutineWD = new CoroutineWithData(Player.main, AssetBundleLoader.LoadUIAsset("chatlog", "PlayerChatCanvas", true));
+            yield return coroutineWD.Coroutine;
+
+            playerChat = ((GameObject)coroutineWD.Result).gameObject.AddComponent<PlayerChat>();
+            yield return playerChat.SetupChatComponents();
```

Comment:
Does this need to be a coroutine?  It introduces a bit of complexity with the new class CoroutineWithData.  Just validating we need the code complexity.

Reply:
We need `CoroutineWithData` if we want to unify the UI load (https://github.com/SubnauticaNitrox/Nitrox/blob/cd644980a009ef29fb7e8d2becd4ae8b5594921e/NitroxClient/Unity/Helper/AssetBundleLoader.cs#L35-L48). The normal coroutine has no option to return a value which is needed in this case. I could remove it and use the lines  for both assets in `PlayerChatManager`.

I think a coroutine of some sort is needed because loading and modifying the asset is a heavy task and shouldn't have the possibly to freeze the game.

Reply:
You can use a callback instead of ienumerator with data class. Ex: Func<TData> as argument to the IEnumerator, invoke it after the yield and then do something with the data at the caller.